server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -227,46 +241,57 @@ void deprecated(final Set<ThreadContext> threadContexts, final String message, f
-                } catch (final IllegalStateException e) {
-                    // ignored; it should be removed shortly
+        threadContextLock.readLock().lock();
+        try{
+            final Iterator<ThreadContext> iterator = threadContexts.iterator();
+            if (iterator.hasNext()) {
+                final String formattedMessage = LoggerMessageFormat.format(message, params);
+                final String warningHeaderValue = formatWarning(formattedMessage);
+                assert WARNING_HEADER_PATTERN.matcher(warningHeaderValue).matches();
+                assert extractWarningValueFromWarningHeader(warningHeaderValue).equals(escapeAndEncode(formattedMessage));
+                while (iterator.hasNext()) {
+                        final ThreadContext next = iterator.next();
+                        next.addResponseHeader("Warning", warningHeaderValue);
                 }
             }
-        }
 
-        if (log) {
-            AccessController.doPrivileged(new PrivilegedAction<Void>() {
-                @SuppressLoggerChecks(reason = "safely delegates to logger")
-                @Override
-                public Void run() {
-                    /**
-                     * There should be only one threadContext (in prod env), @see DeprecationLogger#setThreadContext
-                     */
-                    String opaqueId = getXOpaqueId(threadContexts);
-
-                    logger.warn(new DeprecatedMessage(message, opaqueId, params));
-                    return null;
-                }
-            });
+            if (log) {
+                AccessController.doPrivileged(new PrivilegedAction<Void>() {
+                    @SuppressLoggerChecks(reason = "safely delegates to logger")
+                    @Override
+                    public Void run() {
+                        /**
+                         * There should be only one threadContext (in prod env), @see DeprecationLogger#setThreadContext
+                         */
+                        String opaqueId = getXOpaqueId(threadContexts);
+
+                        logger.warn(new DeprecatedMessage(message, opaqueId, params));
+                        return null;
+                    }
+                });
+            }
+
+        }finally {
+            threadContextLock.readLock().unlock();
         }
+
     }
 
     public String getXOpaqueId(Set<ThreadContext> threadContexts) {
-        return threadContexts.stream()
-                             .filter(t -> t.isClosed() == false)
-                             .filter(t -> t.getHeader(Task.X_OPAQUE_ID) != null)
-                             .findFirst()
-                             .map(t -> t.getHeader(Task.X_OPAQUE_ID))
-                             .orElse("");
+        threadContextLock.readLock().lock();
+        try {
+            for (ThreadContext threadContext : threadContexts) {
+                if (threadContext.isClosed() == false) {
```

Comment:
you can remove the if here since it should never be closed at this point. Maybe add an assert?